pkg/controlplane/controller/legacytokentracking/controller.go

```diff
@@ -0,0 +1,211 @@
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	kubefeatures "k8s.io/kubernetes/pkg/features"
+)
+
+const (
+	configMapName    = "tracked-since"
+	configMapDataKey = "timestamp"
+	freezePeriod     = 30 * time.Minute
+)
+
+// Controller maintains a bool value configmap `tracked-since` in `kube-system`
+// to indicates if the tracking for legacy tokens is enabled in the cluster.
+// For HA clusters, the configmap will be eventually created after all
+// controller instances have enabled the feature. When disabling this feature,
+// existing configmap will be deleted.
+type Controller struct {
+	configMapClient   corev1client.ConfigMapsGetter
+	namespaceClient   corev1client.NamespacesGetter
+	namespaceExists   bool
+	configMapInformer cache.SharedIndexInformer
+	configMapCache    cache.Indexer
+	configMapSynced   cache.InformerSynced
+	queue             workqueue.RateLimitingInterface
+
+	// enabled controls the behavior of the controller: if enabled is true, the
+	// 'tracked-since' configmap will be created; otherwise, the configmap will
+	// be deleted.
+	enabled bool
+	// freezeUntil records the timestamp until when the controller starts to
+	// create the configmap.
+	freezeUntil time.Time
+}
+
+// NewController returns a Controller struct.
+func NewController(cs kubernetes.Interface) *Controller {
+	informer := corev1informers.NewFilteredConfigMapInformer(cs, metav1.NamespaceSystem, 12*time.Hour, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, func(options *metav1.ListOptions) {
+		options.FieldSelector = fields.OneTermEqualSelector("metadata.name", configMapName).String()
+	})
+
+	c := &Controller{
+		configMapClient:   cs.CoreV1(),
+		namespaceClient:   cs.CoreV1(),
+		namespaceExists:   false,
+		queue:             workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "legacy_token_tracking_controller"),
+		configMapInformer: informer,
+		configMapCache:    informer.GetIndexer(),
+		configMapSynced:   informer.HasSynced,
+		enabled:           utilfeature.DefaultFeatureGate.Enabled(kubefeatures.LegacyServiceAccountTokenTracking),
+	}
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			configMap := obj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			configMap := newObj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+		DeleteFunc: func(obj interface{}) {
+			c.freeze()
+			configMap := obj.(*corev1.ConfigMap)
+			c.enqueue(configMap)
+		},
+	})
+
+	return c
+}
+
+func (c *Controller) enqueue(configMap *corev1.ConfigMap) {
+	key, err := cache.MetaNamespaceKeyFunc(configMap)
+	if err != nil {
+		klog.Errorf("Failed to get key for ConfigMap: %v", configMap)
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *Controller) freeze() {
+	// when delete happens and the controller is enabled to create the
```

Comment:
This seems novel. Wouldn't it be easier to add leader election to the controller and simply have the leader decide what the behavior should be?

Reply:
ideally we want a barrier style mechanism where only to create the configmap when all apiservers are enabled and delete immediately when any apiserver is disabled. leaderelection doesn't give that. this is mimicing the barrier.